inventory.lua

```diff
@@ -2,6 +2,43 @@ local S = digilines.S
 
 local pipeworks_enabled = minetest.get_modpath("pipeworks") ~= nil
 
+-- Messages which will be sent in a single batch
+local batched_messages = {}
```

Comment:
This and `last_message_time_for_chest` needs periodic cleanups (e.g. every 30 seconds) in cases where mapblocks are unloaded from memory without sending the message(s).

Reply:
Sorry for late response. I'm not sure that it's the best solution because this means that the message will be lost that might be crucial in some cases and disrupt the logic of the factory. Also, I don't know how likely the situation is when the map region with a digiline chest is unloaded within 0.1 seconds after placing something to there in a batch (it might happen in case of delivering items to there by tubes in large amounts though). But even when it happens, I'm not sure how crucial is that we'll have some extra records in the structure for a while (maybe even until the server is shutting down). Another option would be to create such a timer, and if it detects such a situation, load the area and flush the signal. But in this case, as far as I understand, it'll activate the factory again that might send more items to the chest and trigger another batched message, and we're stuck in a vicious circle. Some players could theoretically even use it to keep their factories running when offline. I currently haven't come up with a solution of this dilemma.

Reply:
Mapblocks can be unloaded from one server step to the next. This case won't occur often, but it can accumulate over time (unless the same mapblocks are loaded again).

**Option 1**: Clearing old entries periodically
```Lua
-- TODO: change to 120 seconds if confirmed working
local cleanup_interval = 30
local function cleanup_batches_from_unloaded_blocks()
	local threshold_time = minetest.get_us_time() - 1E6 * (cleanup_interval + interval_to_batch)
	local pos_to_remove = {}

	for pos_hash, prev_time in pairs(last_message_time_for_chest) do
		if prev_time < threshold_time then
			-- Expired
			table.insert(pos_to_remove, pos_hash) 
		end
	end

	-- Remove all data assigned to this position
	for _, pos_hash in ipairs(pos_to_remove) do
		batched_messages[pos_hash] = nil
		last_message_time_for_chest[pos_hash] = nil
	end

	minetest.after(cleanup_interval, cleanup_batches_from_unloaded_blocks)
end

-- Short time to discover code issues early
minetest.after(5, cleanup_batches_from_unloaded_blocks)
```

**Option 2**: node name validity check 
1. get rid of node timers
2. iterate through `batched_messages` periodically (`minetest.register_globalstep`)
3. process the entry or remove it, depending on whether `minetest.get_node(pos)` returns the digiline node name.

---

Also: might there be nodes that switch between two states (e.g. active/inactive) ? AFAIK when using `minetest.swap_node` or equivalent, the node timer is removed but no `on _destruct` called.

Reply:
It seems I've come up with the perfect solution: use batched messages only for inventory actions triggered by a user. In this case it will be very unlikely that the chest is unloaded after 0.1 seconds after interaction with a user. But even if it happens, it shouldn't be a problem to load the map region again, and flush the batched messages. Regarding actions triggered by tubes, it will be the chest's owner's responsibility to ensure that the incoming (or outcoming) items won't burn the controller.

Reply:
And yes, use `minetest.after()` instead of timers to trigger it

Reply:
Done